// @kyvejs/types/client/kyve/query/v1beta1/funders.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v1.178.0
//   protoc               unknown
// source: kyve/query/v1beta1/funders.proto
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryFundersClientImpl = exports.QueryFundersServiceName = exports.QueryFundingsByPoolResponse = exports.QueryFundingsByPoolRequest = exports.QueryFundingsByFunderResponse = exports.QueryFundingsByFunderRequest = exports.QueryFunderResponse = exports.QueryFunderRequest = exports.QueryFundersResponse = exports.QueryFundersRequest = exports.Funding = exports.FundingStats = exports.Funder = exports.fundingStatusToJSON = exports.fundingStatusFromJSON = exports.FundingStatus = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
var pagination_1 = require("../../../cosmos/base/query/v1beta1/pagination");
var coin_1 = require("../../../cosmos/base/v1beta1/coin");
exports.protobufPackage = "kyve.query.v1beta1";
/** FundingStatus ... */
var FundingStatus;
(function (FundingStatus) {
    /** FUNDING_STATUS_UNSPECIFIED - FundingStatusUnspecified ... */
    FundingStatus[FundingStatus["FUNDING_STATUS_UNSPECIFIED"] = 0] = "FUNDING_STATUS_UNSPECIFIED";
    /** FUNDING_STATUS_ACTIVE - FundingStatusActive status is set when the funding is active. */
    FundingStatus[FundingStatus["FUNDING_STATUS_ACTIVE"] = 1] = "FUNDING_STATUS_ACTIVE";
    /** FUNDING_STATUS_INACTIVE - FundingStatusInactive status is set when the funding has been used up or refunded. */
    FundingStatus[FundingStatus["FUNDING_STATUS_INACTIVE"] = 2] = "FUNDING_STATUS_INACTIVE";
    FundingStatus[FundingStatus["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(FundingStatus = exports.FundingStatus || (exports.FundingStatus = {}));
function fundingStatusFromJSON(object) {
    switch (object) {
        case 0:
        case "FUNDING_STATUS_UNSPECIFIED":
            return FundingStatus.FUNDING_STATUS_UNSPECIFIED;
        case 1:
        case "FUNDING_STATUS_ACTIVE":
            return FundingStatus.FUNDING_STATUS_ACTIVE;
        case 2:
        case "FUNDING_STATUS_INACTIVE":
            return FundingStatus.FUNDING_STATUS_INACTIVE;
        case -1:
        case "UNRECOGNIZED":
        default:
            return FundingStatus.UNRECOGNIZED;
    }
}
exports.fundingStatusFromJSON = fundingStatusFromJSON;
function fundingStatusToJSON(object) {
    switch (object) {
        case FundingStatus.FUNDING_STATUS_UNSPECIFIED:
            return "FUNDING_STATUS_UNSPECIFIED";
        case FundingStatus.FUNDING_STATUS_ACTIVE:
            return "FUNDING_STATUS_ACTIVE";
        case FundingStatus.FUNDING_STATUS_INACTIVE:
            return "FUNDING_STATUS_INACTIVE";
        case FundingStatus.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
exports.fundingStatusToJSON = fundingStatusToJSON;
function createBaseFunder() {
    return { address: "", moniker: "", identity: "", website: "", contact: "", description: "", stats: undefined };
}
exports.Funder = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.address !== "") {
            writer.uint32(10).string(message.address);
        }
        if (message.moniker !== "") {
            writer.uint32(18).string(message.moniker);
        }
        if (message.identity !== "") {
            writer.uint32(26).string(message.identity);
        }
        if (message.website !== "") {
            writer.uint32(34).string(message.website);
        }
        if (message.contact !== "") {
            writer.uint32(42).string(message.contact);
        }
        if (message.description !== "") {
            writer.uint32(50).string(message.description);
        }
        if (message.stats !== undefined) {
            exports.FundingStats.encode(message.stats, writer.uint32(58).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseFunder();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.address = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.moniker = reader.string();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.identity = reader.string();
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.website = reader.string();
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.contact = reader.string();
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.description = reader.string();
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.stats = exports.FundingStats.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            address: isSet(object.address) ? globalThis.String(object.address) : "",
            moniker: isSet(object.moniker) ? globalThis.String(object.moniker) : "",
            identity: isSet(object.identity) ? globalThis.String(object.identity) : "",
            website: isSet(object.website) ? globalThis.String(object.website) : "",
            contact: isSet(object.contact) ? globalThis.String(object.contact) : "",
            description: isSet(object.description) ? globalThis.String(object.description) : "",
            stats: isSet(object.stats) ? exports.FundingStats.fromJSON(object.stats) : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        if (message.address !== "") {
            obj.address = message.address;
        }
        if (message.moniker !== "") {
            obj.moniker = message.moniker;
        }
        if (message.identity !== "") {
            obj.identity = message.identity;
        }
        if (message.website !== "") {
            obj.website = message.website;
        }
        if (message.contact !== "") {
            obj.contact = message.contact;
        }
        if (message.description !== "") {
            obj.description = message.description;
        }
        if (message.stats !== undefined) {
            obj.stats = exports.FundingStats.toJSON(message.stats);
        }
        return obj;
    },
    create: function (base) {
        return exports.Funder.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e, _f;
        var message = createBaseFunder();
        message.address = (_a = object.address) !== null && _a !== void 0 ? _a : "";
        message.moniker = (_b = object.moniker) !== null && _b !== void 0 ? _b : "";
        message.identity = (_c = object.identity) !== null && _c !== void 0 ? _c : "";
        message.website = (_d = object.website) !== null && _d !== void 0 ? _d : "";
        message.contact = (_e = object.contact) !== null && _e !== void 0 ? _e : "";
        message.description = (_f = object.description) !== null && _f !== void 0 ? _f : "";
        message.stats = (object.stats !== undefined && object.stats !== null)
            ? exports.FundingStats.fromPartial(object.stats)
            : undefined;
        return message;
    },
};
function createBaseFundingStats() {
    return { total_used_funds: [], total_allocated_funds: [], total_amount_per_bundle: [], pools_funded: [], score: "0" };
}
exports.FundingStats = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        for (var _i = 0, _a = message.total_used_funds; _i < _a.length; _i++) {
            var v = _a[_i];
            coin_1.Coin.encode(v, writer.uint32(10).fork()).ldelim();
        }
        for (var _b = 0, _c = message.total_allocated_funds; _b < _c.length; _b++) {
            var v = _c[_b];
            coin_1.Coin.encode(v, writer.uint32(18).fork()).ldelim();
        }
        for (var _d = 0, _e = message.total_amount_per_bundle; _d < _e.length; _d++) {
            var v = _e[_d];
            coin_1.Coin.encode(v, writer.uint32(26).fork()).ldelim();
        }
        writer.uint32(34).fork();
        for (var _f = 0, _g = message.pools_funded; _f < _g.length; _f++) {
            var v = _g[_f];
            writer.uint64(v);
        }
        writer.ldelim();
        if (message.score !== "0") {
            writer.uint32(40).uint64(message.score);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseFundingStats();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.total_used_funds.push(coin_1.Coin.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.total_allocated_funds.push(coin_1.Coin.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.total_amount_per_bundle.push(coin_1.Coin.decode(reader, reader.uint32()));
                    continue;
                case 4:
                    if (tag === 32) {
                        message.pools_funded.push(longToString(reader.uint64()));
                        continue;
                    }
                    if (tag === 34) {
                        var end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.pools_funded.push(longToString(reader.uint64()));
                        }
                        continue;
                    }
                    break;
                case 5:
                    if (tag !== 40) {
                        break;
                    }
                    message.score = longToString(reader.uint64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            total_used_funds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.total_used_funds)
                ? object.total_used_funds.map(function (e) { return coin_1.Coin.fromJSON(e); })
                : [],
            total_allocated_funds: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.total_allocated_funds)
                ? object.total_allocated_funds.map(function (e) { return coin_1.Coin.fromJSON(e); })
                : [],
            total_amount_per_bundle: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.total_amount_per_bundle)
                ? object.total_amount_per_bundle.map(function (e) { return coin_1.Coin.fromJSON(e); })
                : [],
            pools_funded: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.pools_funded)
                ? object.pools_funded.map(function (e) { return globalThis.String(e); })
                : [],
            score: isSet(object.score) ? globalThis.String(object.score) : "0",
        };
    },
    toJSON: function (message) {
        var _a, _b, _c, _d;
        var obj = {};
        if ((_a = message.total_used_funds) === null || _a === void 0 ? void 0 : _a.length) {
            obj.total_used_funds = message.total_used_funds.map(function (e) { return coin_1.Coin.toJSON(e); });
        }
        if ((_b = message.total_allocated_funds) === null || _b === void 0 ? void 0 : _b.length) {
            obj.total_allocated_funds = message.total_allocated_funds.map(function (e) { return coin_1.Coin.toJSON(e); });
        }
        if ((_c = message.total_amount_per_bundle) === null || _c === void 0 ? void 0 : _c.length) {
            obj.total_amount_per_bundle = message.total_amount_per_bundle.map(function (e) { return coin_1.Coin.toJSON(e); });
        }
        if ((_d = message.pools_funded) === null || _d === void 0 ? void 0 : _d.length) {
            obj.pools_funded = message.pools_funded;
        }
        if (message.score !== "0") {
            obj.score = message.score;
        }
        return obj;
    },
    create: function (base) {
        return exports.FundingStats.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e;
        var message = createBaseFundingStats();
        message.total_used_funds = ((_a = object.total_used_funds) === null || _a === void 0 ? void 0 : _a.map(function (e) { return coin_1.Coin.fromPartial(e); })) || [];
        message.total_allocated_funds = ((_b = object.total_allocated_funds) === null || _b === void 0 ? void 0 : _b.map(function (e) { return coin_1.Coin.fromPartial(e); })) || [];
        message.total_amount_per_bundle = ((_c = object.total_amount_per_bundle) === null || _c === void 0 ? void 0 : _c.map(function (e) { return coin_1.Coin.fromPartial(e); })) || [];
        message.pools_funded = ((_d = object.pools_funded) === null || _d === void 0 ? void 0 : _d.map(function (e) { return e; })) || [];
        message.score = (_e = object.score) !== null && _e !== void 0 ? _e : "0";
        return message;
    },
};
function createBaseFunding() {
    return { funder_address: "", pool_id: "0", amounts: [], amounts_per_bundle: [], total_funded: [], score: "0" };
}
exports.Funding = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.funder_address !== "") {
            writer.uint32(10).string(message.funder_address);
        }
        if (message.pool_id !== "0") {
            writer.uint32(16).uint64(message.pool_id);
        }
        for (var _i = 0, _a = message.amounts; _i < _a.length; _i++) {
            var v = _a[_i];
            coin_1.Coin.encode(v, writer.uint32(26).fork()).ldelim();
        }
        for (var _b = 0, _c = message.amounts_per_bundle; _b < _c.length; _b++) {
            var v = _c[_b];
            coin_1.Coin.encode(v, writer.uint32(34).fork()).ldelim();
        }
        for (var _d = 0, _e = message.total_funded; _d < _e.length; _d++) {
            var v = _e[_d];
            coin_1.Coin.encode(v, writer.uint32(42).fork()).ldelim();
        }
        if (message.score !== "0") {
            writer.uint32(48).uint64(message.score);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseFunding();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.funder_address = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.pool_id = longToString(reader.uint64());
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.amounts.push(coin_1.Coin.decode(reader, reader.uint32()));
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.amounts_per_bundle.push(coin_1.Coin.decode(reader, reader.uint32()));
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.total_funded.push(coin_1.Coin.decode(reader, reader.uint32()));
                    continue;
                case 6:
                    if (tag !== 48) {
                        break;
                    }
                    message.score = longToString(reader.uint64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            funder_address: isSet(object.funder_address) ? globalThis.String(object.funder_address) : "",
            pool_id: isSet(object.pool_id) ? globalThis.String(object.pool_id) : "0",
            amounts: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.amounts) ? object.amounts.map(function (e) { return coin_1.Coin.fromJSON(e); }) : [],
            amounts_per_bundle: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.amounts_per_bundle)
                ? object.amounts_per_bundle.map(function (e) { return coin_1.Coin.fromJSON(e); })
                : [],
            total_funded: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.total_funded)
                ? object.total_funded.map(function (e) { return coin_1.Coin.fromJSON(e); })
                : [],
            score: isSet(object.score) ? globalThis.String(object.score) : "0",
        };
    },
    toJSON: function (message) {
        var _a, _b, _c;
        var obj = {};
        if (message.funder_address !== "") {
            obj.funder_address = message.funder_address;
        }
        if (message.pool_id !== "0") {
            obj.pool_id = message.pool_id;
        }
        if ((_a = message.amounts) === null || _a === void 0 ? void 0 : _a.length) {
            obj.amounts = message.amounts.map(function (e) { return coin_1.Coin.toJSON(e); });
        }
        if ((_b = message.amounts_per_bundle) === null || _b === void 0 ? void 0 : _b.length) {
            obj.amounts_per_bundle = message.amounts_per_bundle.map(function (e) { return coin_1.Coin.toJSON(e); });
        }
        if ((_c = message.total_funded) === null || _c === void 0 ? void 0 : _c.length) {
            obj.total_funded = message.total_funded.map(function (e) { return coin_1.Coin.toJSON(e); });
        }
        if (message.score !== "0") {
            obj.score = message.score;
        }
        return obj;
    },
    create: function (base) {
        return exports.Funding.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d, _e, _f;
        var message = createBaseFunding();
        message.funder_address = (_a = object.funder_address) !== null && _a !== void 0 ? _a : "";
        message.pool_id = (_b = object.pool_id) !== null && _b !== void 0 ? _b : "0";
        message.amounts = ((_c = object.amounts) === null || _c === void 0 ? void 0 : _c.map(function (e) { return coin_1.Coin.fromPartial(e); })) || [];
        message.amounts_per_bundle = ((_d = object.amounts_per_bundle) === null || _d === void 0 ? void 0 : _d.map(function (e) { return coin_1.Coin.fromPartial(e); })) || [];
        message.total_funded = ((_e = object.total_funded) === null || _e === void 0 ? void 0 : _e.map(function (e) { return coin_1.Coin.fromPartial(e); })) || [];
        message.score = (_f = object.score) !== null && _f !== void 0 ? _f : "0";
        return message;
    },
};
function createBaseQueryFundersRequest() {
    return { pagination: undefined, search: "" };
}
exports.QueryFundersRequest = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(10).fork()).ldelim();
        }
        if (message.search !== "") {
            writer.uint32(18).string(message.search);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseQueryFundersRequest();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.search = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            pagination: isSet(object.pagination) ? pagination_1.PageRequest.fromJSON(object.pagination) : undefined,
            search: isSet(object.search) ? globalThis.String(object.search) : "",
        };
    },
    toJSON: function (message) {
        var obj = {};
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        if (message.search !== "") {
            obj.search = message.search;
        }
        return obj;
    },
    create: function (base) {
        return exports.QueryFundersRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseQueryFundersRequest();
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageRequest.fromPartial(object.pagination)
            : undefined;
        message.search = (_a = object.search) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseQueryFundersResponse() {
    return { pagination: undefined, funders: [] };
}
exports.QueryFundersResponse = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(10).fork()).ldelim();
        }
        for (var _i = 0, _a = message.funders; _i < _a.length; _i++) {
            var v = _a[_i];
            exports.Funder.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseQueryFundersResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.funders.push(exports.Funder.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            pagination: isSet(object.pagination) ? pagination_1.PageResponse.fromJSON(object.pagination) : undefined,
            funders: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.funders) ? object.funders.map(function (e) { return exports.Funder.fromJSON(e); }) : [],
        };
    },
    toJSON: function (message) {
        var _a;
        var obj = {};
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        if ((_a = message.funders) === null || _a === void 0 ? void 0 : _a.length) {
            obj.funders = message.funders.map(function (e) { return exports.Funder.toJSON(e); });
        }
        return obj;
    },
    create: function (base) {
        return exports.QueryFundersResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseQueryFundersResponse();
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageResponse.fromPartial(object.pagination)
            : undefined;
        message.funders = ((_a = object.funders) === null || _a === void 0 ? void 0 : _a.map(function (e) { return exports.Funder.fromPartial(e); })) || [];
        return message;
    },
};
function createBaseQueryFunderRequest() {
    return { address: "", status: 0 };
}
exports.QueryFunderRequest = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.address !== "") {
            writer.uint32(10).string(message.address);
        }
        if (message.status !== 0) {
            writer.uint32(24).int32(message.status);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseQueryFunderRequest();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.address = reader.string();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.status = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            address: isSet(object.address) ? globalThis.String(object.address) : "",
            status: isSet(object.status) ? fundingStatusFromJSON(object.status) : 0,
        };
    },
    toJSON: function (message) {
        var obj = {};
        if (message.address !== "") {
            obj.address = message.address;
        }
        if (message.status !== 0) {
            obj.status = fundingStatusToJSON(message.status);
        }
        return obj;
    },
    create: function (base) {
        return exports.QueryFunderRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseQueryFunderRequest();
        message.address = (_a = object.address) !== null && _a !== void 0 ? _a : "";
        message.status = (_b = object.status) !== null && _b !== void 0 ? _b : 0;
        return message;
    },
};
function createBaseQueryFunderResponse() {
    return { funder: undefined, fundings: [] };
}
exports.QueryFunderResponse = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.funder !== undefined) {
            exports.Funder.encode(message.funder, writer.uint32(10).fork()).ldelim();
        }
        for (var _i = 0, _a = message.fundings; _i < _a.length; _i++) {
            var v = _a[_i];
            exports.Funding.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseQueryFunderResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.funder = exports.Funder.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.fundings.push(exports.Funding.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            funder: isSet(object.funder) ? exports.Funder.fromJSON(object.funder) : undefined,
            fundings: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.fundings) ? object.fundings.map(function (e) { return exports.Funding.fromJSON(e); }) : [],
        };
    },
    toJSON: function (message) {
        var _a;
        var obj = {};
        if (message.funder !== undefined) {
            obj.funder = exports.Funder.toJSON(message.funder);
        }
        if ((_a = message.fundings) === null || _a === void 0 ? void 0 : _a.length) {
            obj.fundings = message.fundings.map(function (e) { return exports.Funding.toJSON(e); });
        }
        return obj;
    },
    create: function (base) {
        return exports.QueryFunderResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseQueryFunderResponse();
        message.funder = (object.funder !== undefined && object.funder !== null)
            ? exports.Funder.fromPartial(object.funder)
            : undefined;
        message.fundings = ((_a = object.fundings) === null || _a === void 0 ? void 0 : _a.map(function (e) { return exports.Funding.fromPartial(e); })) || [];
        return message;
    },
};
function createBaseQueryFundingsByFunderRequest() {
    return { pagination: undefined, address: "", status: 0 };
}
exports.QueryFundingsByFunderRequest = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(10).fork()).ldelim();
        }
        if (message.address !== "") {
            writer.uint32(18).string(message.address);
        }
        if (message.status !== 0) {
            writer.uint32(24).int32(message.status);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseQueryFundingsByFunderRequest();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.address = reader.string();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.status = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            pagination: isSet(object.pagination) ? pagination_1.PageRequest.fromJSON(object.pagination) : undefined,
            address: isSet(object.address) ? globalThis.String(object.address) : "",
            status: isSet(object.status) ? fundingStatusFromJSON(object.status) : 0,
        };
    },
    toJSON: function (message) {
        var obj = {};
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        if (message.address !== "") {
            obj.address = message.address;
        }
        if (message.status !== 0) {
            obj.status = fundingStatusToJSON(message.status);
        }
        return obj;
    },
    create: function (base) {
        return exports.QueryFundingsByFunderRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseQueryFundingsByFunderRequest();
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageRequest.fromPartial(object.pagination)
            : undefined;
        message.address = (_a = object.address) !== null && _a !== void 0 ? _a : "";
        message.status = (_b = object.status) !== null && _b !== void 0 ? _b : 0;
        return message;
    },
};
function createBaseQueryFundingsByFunderResponse() {
    return { pagination: undefined, fundings: [] };
}
exports.QueryFundingsByFunderResponse = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(10).fork()).ldelim();
        }
        for (var _i = 0, _a = message.fundings; _i < _a.length; _i++) {
            var v = _a[_i];
            exports.Funding.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseQueryFundingsByFunderResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.fundings.push(exports.Funding.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            pagination: isSet(object.pagination) ? pagination_1.PageResponse.fromJSON(object.pagination) : undefined,
            fundings: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.fundings) ? object.fundings.map(function (e) { return exports.Funding.fromJSON(e); }) : [],
        };
    },
    toJSON: function (message) {
        var _a;
        var obj = {};
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        if ((_a = message.fundings) === null || _a === void 0 ? void 0 : _a.length) {
            obj.fundings = message.fundings.map(function (e) { return exports.Funding.toJSON(e); });
        }
        return obj;
    },
    create: function (base) {
        return exports.QueryFundingsByFunderResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseQueryFundingsByFunderResponse();
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageResponse.fromPartial(object.pagination)
            : undefined;
        message.fundings = ((_a = object.fundings) === null || _a === void 0 ? void 0 : _a.map(function (e) { return exports.Funding.fromPartial(e); })) || [];
        return message;
    },
};
function createBaseQueryFundingsByPoolRequest() {
    return { pagination: undefined, pool_id: "0", status: 0 };
}
exports.QueryFundingsByPoolRequest = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.pagination !== undefined) {
            pagination_1.PageRequest.encode(message.pagination, writer.uint32(10).fork()).ldelim();
        }
        if (message.pool_id !== "0") {
            writer.uint32(16).uint64(message.pool_id);
        }
        if (message.status !== 0) {
            writer.uint32(24).int32(message.status);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseQueryFundingsByPoolRequest();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.pagination = pagination_1.PageRequest.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.pool_id = longToString(reader.uint64());
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.status = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            pagination: isSet(object.pagination) ? pagination_1.PageRequest.fromJSON(object.pagination) : undefined,
            pool_id: isSet(object.pool_id) ? globalThis.String(object.pool_id) : "0",
            status: isSet(object.status) ? fundingStatusFromJSON(object.status) : 0,
        };
    },
    toJSON: function (message) {
        var obj = {};
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageRequest.toJSON(message.pagination);
        }
        if (message.pool_id !== "0") {
            obj.pool_id = message.pool_id;
        }
        if (message.status !== 0) {
            obj.status = fundingStatusToJSON(message.status);
        }
        return obj;
    },
    create: function (base) {
        return exports.QueryFundingsByPoolRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseQueryFundingsByPoolRequest();
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageRequest.fromPartial(object.pagination)
            : undefined;
        message.pool_id = (_a = object.pool_id) !== null && _a !== void 0 ? _a : "0";
        message.status = (_b = object.status) !== null && _b !== void 0 ? _b : 0;
        return message;
    },
};
function createBaseQueryFundingsByPoolResponse() {
    return { pagination: undefined, fundings: [] };
}
exports.QueryFundingsByPoolResponse = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.pagination !== undefined) {
            pagination_1.PageResponse.encode(message.pagination, writer.uint32(10).fork()).ldelim();
        }
        for (var _i = 0, _a = message.fundings; _i < _a.length; _i++) {
            var v = _a[_i];
            exports.Funding.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseQueryFundingsByPoolResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.pagination = pagination_1.PageResponse.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.fundings.push(exports.Funding.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            pagination: isSet(object.pagination) ? pagination_1.PageResponse.fromJSON(object.pagination) : undefined,
            fundings: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.fundings) ? object.fundings.map(function (e) { return exports.Funding.fromJSON(e); }) : [],
        };
    },
    toJSON: function (message) {
        var _a;
        var obj = {};
        if (message.pagination !== undefined) {
            obj.pagination = pagination_1.PageResponse.toJSON(message.pagination);
        }
        if ((_a = message.fundings) === null || _a === void 0 ? void 0 : _a.length) {
            obj.fundings = message.fundings.map(function (e) { return exports.Funding.toJSON(e); });
        }
        return obj;
    },
    create: function (base) {
        return exports.QueryFundingsByPoolResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseQueryFundingsByPoolResponse();
        message.pagination = (object.pagination !== undefined && object.pagination !== null)
            ? pagination_1.PageResponse.fromPartial(object.pagination)
            : undefined;
        message.fundings = ((_a = object.fundings) === null || _a === void 0 ? void 0 : _a.map(function (e) { return exports.Funding.fromPartial(e); })) || [];
        return message;
    },
};
exports.QueryFundersServiceName = "kyve.query.v1beta1.QueryFunders";
var QueryFundersClientImpl = /** @class */ (function () {
    function QueryFundersClientImpl(rpc, opts) {
        this.service = (opts === null || opts === void 0 ? void 0 : opts.service) || exports.QueryFundersServiceName;
        this.rpc = rpc;
        this.Funders = this.Funders.bind(this);
        this.Funder = this.Funder.bind(this);
        this.FundingsByFunder = this.FundingsByFunder.bind(this);
        this.FundingsByPool = this.FundingsByPool.bind(this);
    }
    QueryFundersClientImpl.prototype.Funders = function (request) {
        var data = exports.QueryFundersRequest.encode(request).finish();
        var promise = this.rpc.request(this.service, "Funders", data);
        return promise.then(function (data) { return exports.QueryFundersResponse.decode(minimal_1.default.Reader.create(data)); });
    };
    QueryFundersClientImpl.prototype.Funder = function (request) {
        var data = exports.QueryFunderRequest.encode(request).finish();
        var promise = this.rpc.request(this.service, "Funder", data);
        return promise.then(function (data) { return exports.QueryFunderResponse.decode(minimal_1.default.Reader.create(data)); });
    };
    QueryFundersClientImpl.prototype.FundingsByFunder = function (request) {
        var data = exports.QueryFundingsByFunderRequest.encode(request).finish();
        var promise = this.rpc.request(this.service, "FundingsByFunder", data);
        return promise.then(function (data) { return exports.QueryFundingsByFunderResponse.decode(minimal_1.default.Reader.create(data)); });
    };
    QueryFundersClientImpl.prototype.FundingsByPool = function (request) {
        var data = exports.QueryFundingsByPoolRequest.encode(request).finish();
        var promise = this.rpc.request(this.service, "FundingsByPool", data);
        return promise.then(function (data) { return exports.QueryFundingsByPoolResponse.decode(minimal_1.default.Reader.create(data)); });
    };
    return QueryFundersClientImpl;
}());
exports.QueryFundersClientImpl = QueryFundersClientImpl;
function longToString(long) {
    return long.toString();
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
