// @ardrive/turbo-sdk/lib/cjs/common/token/arweave.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ARToTokenAmount = exports.WinstonToTokenAmount = exports.ArweaveToken = void 0;
/**
 * Copyright (C) 2022-2024 Permanent Data Solutions, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const arweave_1 = __importDefault(require("arweave"));
const bignumber_js_1 = require("bignumber.js");
const node_buffer_1 = require("node:buffer");
const base64_js_1 = require("../../utils/base64.js");
const common_js_1 = require("../../utils/common.js");
const logger_js_1 = require("../logger.js");
const ArweaveClass = 
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore -- Access the correct class constructor for Arweave
arweave_1.default.default?.default || arweave_1.default.default || arweave_1.default;
class ArweaveToken {
    constructor({ gatewayUrl = 'https://arweave.net', arweave, logger = logger_js_1.TurboWinstonLogger.default, mintU = true, pollingOptions = {
        maxAttempts: 10,
        pollingIntervalMs: 3_000,
        initialBackoffMs: 7_000,
    }, } = {}) {
        const url = new URL(gatewayUrl);
        this.arweave =
            arweave ??
                new ArweaveClass({
                    host: url.hostname,
                    port: url.port,
                    protocol: url.protocol.replace(':', ''),
                });
        this.logger = logger;
        this.mintU = mintU;
        this.pollingOptions = pollingOptions;
    }
    async createAndSubmitTx({ feeMultiplier, target, tokenAmount, signer, }) {
        const tx = await this.arweave.createTransaction({
            target,
            quantity: tokenAmount.toString(),
            data: '',
        });
        if (feeMultiplier !== 1) {
            tx.reward = (0, bignumber_js_1.BigNumber)(tx.reward)
                .times((0, bignumber_js_1.BigNumber)(feeMultiplier))
                .toFixed(0, bignumber_js_1.BigNumber.ROUND_UP);
        }
        if (this.mintU) {
            tx.addTag('App-Name', 'SmartWeaveAction');
            tx.addTag('App-Version', '0.3.0'); // cspell:disable
            tx.addTag('Contract', 'KTzTXT_ANmF84fWEKHzWURD1LWd9QaFR9yfYUwH2Lxw'); // cspell:enable
            tx.addTag('Input', JSON.stringify({ function: 'mint' }));
        }
        const publicKeyB64Url = (0, base64_js_1.toB64Url)(await signer.getPublicKey());
        tx.setOwner(publicKeyB64Url);
        const dataToSign = await tx.getSignatureData();
        const signatureUint8Array = await signer.signData(dataToSign);
        const signatureBuffer = node_buffer_1.Buffer.from(signatureUint8Array);
        const id = (0, base64_js_1.sha256B64Url)(signatureBuffer);
        tx.setSignature({
            id: id,
            owner: publicKeyB64Url,
            signature: (0, base64_js_1.toB64Url)(signatureBuffer),
        });
        this.logger.debug('Submitting fund transaction...', { id });
        await this.submitTx(tx);
        return { id, target, reward: tx.reward };
    }
    async pollForTxBeingAvailable({ txId, }) {
        const { maxAttempts, pollingIntervalMs, initialBackoffMs } = this.pollingOptions;
        this.logger.debug('Polling for transaction...', { txId });
        await (0, common_js_1.sleep)(initialBackoffMs);
        let attempts = 0;
        while (attempts < maxAttempts) {
            let transaction;
            attempts++;
            try {
                const response = await this.arweave.api.post('/graphql', {
                    query: `
          query {
            transaction(id: "${txId}") {
              recipient
              owner {
                address
              }
              quantity {
                winston
              }
            }
          }
          `,
                });
                transaction = response?.data?.data?.transaction;
            }
            catch (err) {
                // Continue retries when request errors
                this.logger.debug('Failed to poll for transaction...', { err });
            }
            if (transaction) {
                return;
            }
            this.logger.debug('Transaction not found...', {
                txId,
                attempts,
                maxAttempts,
                pollingIntervalMs,
            });
            await (0, common_js_1.sleep)(pollingIntervalMs);
        }
        throw new Error('Transaction not found after polling, transaction id: ' + txId);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    async submitTx(tx) {
        try {
            const response = await this.arweave.transactions.post(tx);
            if (response.status !== 200) {
                throw new Error('Failed to post transaction -- ' +
                    `Status ${response.status}, ${response.statusText}, ${response.data}`);
            }
            this.logger.debug('Successfully posted fund transaction...', { tx });
        }
        catch (err) {
            throw new Error(`Failed to post transaction -- ${err instanceof Error ? err.message : err}`);
        }
        this.logger.debug('Posted transaction...', { tx });
    }
}
exports.ArweaveToken = ArweaveToken;
const WinstonToTokenAmount = (winston) => winston;
exports.WinstonToTokenAmount = WinstonToTokenAmount;
const ARToTokenAmount = (ar) => new bignumber_js_1.BigNumber(ar).times(1e12).valueOf();
exports.ARToTokenAmount = ARToTokenAmount;
