// @keplr-wallet/proto-types/proto-types-gen/src/secret/compute/v1beta1/params.ts
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal";

export const protobufPackage = "secret.compute.v1beta1";

/** Params defines the parameters for the compute module. */
export interface Params {
  /**
   * CompileCost is how much SDK gas we charge *per byte* for compiling WASM
   * code.
   */
  compileCost: string;
  /** MaxContractSize is the maximum size of contract to store in bytes. */
  maxContractSize: string;
}

function createBaseParams(): Params {
  return { compileCost: "", maxContractSize: "0" };
}

export const Params = {
  encode(
    message: Params,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    if (message.compileCost !== "") {
      writer.uint32(10).string(message.compileCost);
    }
    if (message.maxContractSize !== "0") {
      writer.uint32(16).uint64(message.maxContractSize);
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): Params {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseParams();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.compileCost = reader.string();
          break;
        case 2:
          message.maxContractSize = longToString(reader.uint64() as Long);
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): Params {
    return {
      compileCost: isSet(object.compileCost) ? String(object.compileCost) : "",
      maxContractSize: isSet(object.maxContractSize)
        ? String(object.maxContractSize)
        : "0",
    };
  },

  toJSON(message: Params): unknown {
    const obj: any = {};
    message.compileCost !== undefined &&
      (obj.compileCost = message.compileCost);
    message.maxContractSize !== undefined &&
      (obj.maxContractSize = message.maxContractSize);
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<Params>, I>>(object: I): Params {
    const message = createBaseParams();
    message.compileCost = object.compileCost ?? "";
    message.maxContractSize = object.maxContractSize ?? "0";
    return message;
  },
};

type Builtin =
  | Date
  | Function
  | Uint8Array
  | string
  | number
  | boolean
  | undefined;

export type DeepPartial<T> = T extends Builtin
  ? T
  : T extends Array<infer U>
  ? Array<DeepPartial<U>>
  : T extends ReadonlyArray<infer U>
  ? ReadonlyArray<DeepPartial<U>>
  : T extends {}
  ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin
  ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & Record<
        Exclude<keyof I, KeysOfUnion<P>>,
        never
      >;

function longToString(long: Long) {
  return long.toString();
}

if (_m0.util.Long !== Long) {
  _m0.util.Long = Long as any;
  _m0.configure();
}

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
