// @keplr-wallet/proto-types/stride/stakeibc/ica_account.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.iCAAccountTypeToJSON = exports.iCAAccountTypeFromJSON = exports.ICAAccountType = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "stride.stakeibc";
var ICAAccountType;
(function (ICAAccountType) {
    ICAAccountType[ICAAccountType["DELEGATION"] = 0] = "DELEGATION";
    ICAAccountType[ICAAccountType["FEE"] = 1] = "FEE";
    ICAAccountType[ICAAccountType["WITHDRAWAL"] = 2] = "WITHDRAWAL";
    ICAAccountType[ICAAccountType["REDEMPTION"] = 3] = "REDEMPTION";
    ICAAccountType[ICAAccountType["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(ICAAccountType = exports.ICAAccountType || (exports.ICAAccountType = {}));
function iCAAccountTypeFromJSON(object) {
    switch (object) {
        case 0:
        case "DELEGATION":
            return ICAAccountType.DELEGATION;
        case 1:
        case "FEE":
            return ICAAccountType.FEE;
        case 2:
        case "WITHDRAWAL":
            return ICAAccountType.WITHDRAWAL;
        case 3:
        case "REDEMPTION":
            return ICAAccountType.REDEMPTION;
        case -1:
        case "UNRECOGNIZED":
        default:
            return ICAAccountType.UNRECOGNIZED;
    }
}
exports.iCAAccountTypeFromJSON = iCAAccountTypeFromJSON;
function iCAAccountTypeToJSON(object) {
    switch (object) {
        case ICAAccountType.DELEGATION:
            return "DELEGATION";
        case ICAAccountType.FEE:
            return "FEE";
        case ICAAccountType.WITHDRAWAL:
            return "WITHDRAWAL";
        case ICAAccountType.REDEMPTION:
            return "REDEMPTION";
        default:
            return "UNKNOWN";
    }
}
exports.iCAAccountTypeToJSON = iCAAccountTypeToJSON;
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
