// @keplr-wallet/proto-types/secret/compute/v1beta1/params.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "secret.compute.v1beta1";
function createBaseParams() {
    return { compileCost: "", maxContractSize: "0" };
}
exports.Params = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.compileCost !== "") {
            writer.uint32(10).string(message.compileCost);
        }
        if (message.maxContractSize !== "0") {
            writer.uint32(16).uint64(message.maxContractSize);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : new minimal_1.default.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.compileCost = reader.string();
                    break;
                case 2:
                    message.maxContractSize = longToString(reader.uint64());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        return {
            compileCost: isSet(object.compileCost) ? String(object.compileCost) : "",
            maxContractSize: isSet(object.maxContractSize)
                ? String(object.maxContractSize)
                : "0",
        };
    },
    toJSON(message) {
        const obj = {};
        message.compileCost !== undefined &&
            (obj.compileCost = message.compileCost);
        message.maxContractSize !== undefined &&
            (obj.maxContractSize = message.maxContractSize);
        return obj;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseParams();
        message.compileCost = (_a = object.compileCost) !== null && _a !== void 0 ? _a : "";
        message.maxContractSize = (_b = object.maxContractSize) !== null && _b !== void 0 ? _b : "0";
        return message;
    },
};
function longToString(long) {
    return long.toString();
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
