// @kyvejs/types/lcd/kyve/bundles/v1beta1/params.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v1.178.0
//   protoc               unknown
// source: kyve/bundles/v1beta1/params.proto
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.StorageCost = exports.protobufPackage = void 0;
/* eslint-disable */
var long_1 = __importDefault(require("long"));
var minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "kyve.bundles.v1beta1";
function createBaseStorageCost() {
    return { storage_provider_id: 0, cost: "" };
}
exports.StorageCost = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.storage_provider_id !== 0) {
            writer.uint32(8).uint32(message.storage_provider_id);
        }
        if (message.cost !== "") {
            writer.uint32(18).string(message.cost);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseStorageCost();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.storage_provider_id = reader.uint32();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.cost = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            storage_provider_id: isSet(object.storage_provider_id) ? globalThis.Number(object.storage_provider_id) : 0,
            cost: isSet(object.cost) ? globalThis.String(object.cost) : "",
        };
    },
    toJSON: function (message) {
        var obj = {};
        if (message.storage_provider_id !== 0) {
            obj.storage_provider_id = Math.round(message.storage_provider_id);
        }
        if (message.cost !== "") {
            obj.cost = message.cost;
        }
        return obj;
    },
    create: function (base) {
        return exports.StorageCost.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseStorageCost();
        message.storage_provider_id = (_a = object.storage_provider_id) !== null && _a !== void 0 ? _a : 0;
        message.cost = (_b = object.cost) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseParams() {
    return { upload_timeout: "0", storage_costs: [], network_fee: "", max_points: "0" };
}
exports.Params = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_1.default.Writer.create(); }
        if (message.upload_timeout !== "0") {
            writer.uint32(8).uint64(message.upload_timeout);
        }
        for (var _i = 0, _a = message.storage_costs; _i < _a.length; _i++) {
            var v = _a[_i];
            exports.StorageCost.encode(v, writer.uint32(18).fork()).ldelim();
        }
        if (message.network_fee !== "") {
            writer.uint32(26).string(message.network_fee);
        }
        if (message.max_points !== "0") {
            writer.uint32(32).uint64(message.max_points);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseParams();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.upload_timeout = longToString(reader.uint64());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.storage_costs.push(exports.StorageCost.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.network_fee = reader.string();
                    continue;
                case 4:
                    if (tag !== 32) {
                        break;
                    }
                    message.max_points = longToString(reader.uint64());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            upload_timeout: isSet(object.upload_timeout) ? globalThis.String(object.upload_timeout) : "0",
            storage_costs: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.storage_costs)
                ? object.storage_costs.map(function (e) { return exports.StorageCost.fromJSON(e); })
                : [],
            network_fee: isSet(object.network_fee) ? globalThis.String(object.network_fee) : "",
            max_points: isSet(object.max_points) ? globalThis.String(object.max_points) : "0",
        };
    },
    toJSON: function (message) {
        var _a;
        var obj = {};
        if (message.upload_timeout !== "0") {
            obj.upload_timeout = message.upload_timeout;
        }
        if ((_a = message.storage_costs) === null || _a === void 0 ? void 0 : _a.length) {
            obj.storage_costs = message.storage_costs.map(function (e) { return exports.StorageCost.toJSON(e); });
        }
        if (message.network_fee !== "") {
            obj.network_fee = message.network_fee;
        }
        if (message.max_points !== "0") {
            obj.max_points = message.max_points;
        }
        return obj;
    },
    create: function (base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d;
        var message = createBaseParams();
        message.upload_timeout = (_a = object.upload_timeout) !== null && _a !== void 0 ? _a : "0";
        message.storage_costs = ((_b = object.storage_costs) === null || _b === void 0 ? void 0 : _b.map(function (e) { return exports.StorageCost.fromPartial(e); })) || [];
        message.network_fee = (_c = object.network_fee) !== null && _c !== void 0 ? _c : "";
        message.max_points = (_d = object.max_points) !== null && _d !== void 0 ? _d : "0";
        return message;
    },
};
function longToString(long) {
    return long.toString();
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
